File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.go
func setZeroes(matrix [][]int) {
    firstRow, firstCol := false, false
    rows, cols := len(matrix), len(matrix[0])

    for i := 0; i < rows; i++ {
        if matrix[i][0] == 0 {
            firstCol = true
        }
    }

    for j := 0; j < cols; j++ {
        if matrix[0][j] == 0 {
            firstRow = true
        }
    }

    for i := 1; i < rows; i++ {
        for j := 1; j < cols; j++ {
            if matrix[i][j] == 0 {
                matrix[i][0] = 0
                matrix[0][j] = 0
            }
        }
    }

    for i := 1; i < rows; i++ {
        for j := 1; j < cols; j++ {
            if matrix[i][0] == 0 || matrix[0][j] == 0 {
                matrix[i][j] = 0
            }
        }
    }

    if firstRow {
        for j := 0; j < cols; j++ {
            matrix[0][j] = 0
        }
    }

    if firstCol {
        for i := 0; i < rows; i++ {
            matrix[i][0] = 0
        }
    }
}

File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.py
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        firstrow = False
        firstcol = False
        
        for i in range(len(matrix)):
            if matrix[i][0] == 0:
                firstcol = True

        for j in range(len(matrix[0])):
            if matrix[0][j] == 0:
                firstrow = True

        for i in range(1, len(matrix)):
            for j in range(1, len(matrix[0])):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0
                    matrix[0][j] = 0

        for i in range(1, len(matrix)):
            for j in range(1, len(matrix[0])):
                if matrix[i][0] == 0 or matrix[0][j] == 0:
                    matrix[i][j] = 0

        if firstrow:
            for j in range(len(matrix[0])):
                matrix[0][j] = 0

        if firstcol:
            for i in range(len(matrix)):
                matrix[i][0] = 0
